src/Wallet.ts

```diff
@@ -0,0 +1,252 @@
+  addresses: AddressMetadata[];
+  changeAddresses: AddressMetadata[];
+  addressesWithUtxo: AddressMetadata[];
+};
+
+export type UtxoWithMetadata = Utxo & {
+  address: string;
+  derivationPath: string;
+};
+
+export type AddressCollectionStats = Omit<
+  WalletStats,
+  'nextUnusedChangeAddress' | 'changeAddresses'
+>;
+
+export class Wallet {
+  constructor(
+    private blockchainService: BlockchainService,
+    private gapLimit = 20,
+    private derivationPath = "m/84'/0'/0'",
+  ) {}
+
+  generateMnemonic = () => bip39.generateMnemonic();
+
+  private createHdRoot = async (mnemonic: string) => {
+    const seed = await bip39.mnemonicToSeed(mnemonic);
+    const hdRoot = bip32.fromSeed(seed);
+
+    return hdRoot;
+  };
+
+  getMasterPublicKey = async (mnemonic: string) => {
+    const hdRoot = await this.createHdRoot(mnemonic);
+    const derivedNode = hdRoot.derivePath(this.derivationPath).neutered();
+    const masterPublicKey = derivedNode.toBase58();
+
+    return masterPublicKey;
+  };
+
+  private fetchAddressCollectionStats = async (
+    node: bip32.BIP32Interface,
+  ): Promise<AddressCollectionStats> => {
+    let gap = 0;
+    let addressIndex = 0;
+    let balance = 0;
+    let pendingBalance = 0;
+    let addresses: AddressMetadata[] = [];
+    let addressesWithUtxo: AddressMetadata[] = [];
+    let nextUnusedAddress: AddressMetadata | null = null;
+    while (gap < this.gapLimit) {
+      const childNode = node.derive(addressIndex);
+      const { address } = bitcoinjs.payments.p2wpkh({
+        pubkey: childNode.publicKey,
+      });
+
+      const addressStats = await this.blockchainService.fetchAddressStats(
+        address!,
+      );
+
+      const { chain_stats, mempool_stats } = addressStats;
+
+      const isUsedAddress =
+        chain_stats.funded_txo_count > 0 ||
+        chain_stats.spent_txo_count > 0 ||
+        mempool_stats.funded_txo_count > 0 ||
+        mempool_stats.spent_txo_count > 0;
+
+      const addressBalance =
+        chain_stats.funded_txo_sum - chain_stats.spent_txo_sum;
+
+      balance += addressBalance;
+      pendingBalance +=
+        mempool_stats.funded_txo_sum - mempool_stats.spent_txo_sum;
+
+      const addressMetadata = {
+        address: address!,
+        derivationPath: `${node.index}/${addressIndex}`,
+        publicKey: childNode.publicKey,
+      };
+
+      addresses.push(addressMetadata);
+
+      if (addressBalance > 0) {
+        addressesWithUtxo.push(addressMetadata);
+      }
+
+      if (!isUsedAddress && !nextUnusedAddress) {
+        nextUnusedAddress = addressMetadata;
+      }
+
+      if (!isUsedAddress) {
+        gap++;
+      } else {
+        gap = 0;
+      }
+
+      addressIndex++;
+    }
+
+    return {
+      balance,
+      pendingBalance,
+      addresses,
+      addressesWithUtxo,
+      nextUnusedAddress: nextUnusedAddress!,
+    };
+  };
+
+  fetchWalletStats = async (masterPublicKey: string): Promise<WalletStats> => {
```

Comment:
The user is expected to call this function to retrieve complete stats of the Wallet. The stats include:

* Total wallet balance
* Total wallet pending balance (mempool transactions)
* A list of Wallet addresses (up to the configured gap limit)
* A list of Wallet change addresses (up to the configured limit)
* A list of addresses which include UTXOs
* The next unused address
* The next unused change address